src/Static.jl

```diff
@@ -939,4 +930,39 @@ function Base.show(io::IO, m::MIME"text/plain", @nospecialize(x::NDIndex))
     show(io, m, Tuple(x))
 end
 
+# field and property accessors
+"""
+    field_type(::Type{T}, f)
+
+Functionally equivalent to `fieldtype(T, f)` except `f` may be a static type.
+"""
+@inline field_type(T::Type, f::Union{Int, Symbol}) = fieldtype(T, f)
+@inline field_type(::Type{T}, ::StaticInt{N}) where {T, N} = fieldtype(T, N)
+@inline field_type(::Type{T}, ::StaticSymbol{S}) where {T, S} = fieldtype(T, S)
+
+function (::Base.Fix2{typeof(getfield), <:Union{StaticSymbol{f}, StaticInt{f}}})(x) where {f
+                                                                                           }
+    getfield(x, f)
+end
+function (::Base.Fix2{typeof(fieldtype), <:Union{StaticSymbol{f}, StaticInt{f}}})(x) where {
+                                                                                            f
+                                                                                            }
+    fieldtype(x, f)
+end
+
+Base.getproperty(x, ::StaticSymbol{S}) where {S} = getproperty(x, S)
+Base.setproperty!(x, ::StaticSymbol{S}, v) where {S} = setproperty!(x, S, v)
+Base.hasproperty(x, ::StaticSymbol{S}) where {S} = hasproperty(x, S)
+
+Base.getindex(nt::NamedTuple, ::StaticSymbol{S}) where {S} = getfield(nt, S)
+function Base.getindex(nt::NamedTuple, idxs::Tuple{<:StaticSymbol, Vararg{<:StaticSymbol}})
+    NamedTuple{known(idxs)}(nt)
+end
+function Base.setindex(nt::NamedTuple, v, ::StaticSymbol{S}) where {S}
+    merge(nt, NamedTuple{(S,)}((v,)))
```

Comment:
For named tuples, `Base.merge` is not type-stable for nested structures. It uses `if @generated`, which seems to allow the compiler to give up very early on type inference. 

I've solved this in my packages by defining an alternate function to use instead of `merge` that's always generated (so no type-unstable fallback). I'd file an issue about this, but I so often see compiler speed favored over strong type inference, so I'm not sure it would be well-received.

Reply:
Do you have a mwe example where this fails so I can have a test that avoids this?

Reply:
Also worth considering: Should this be 
```julia
function Base.setindex(nt::NamedTuple{N,T}, v, ::StaticSymbol{S}) where {S,N,T}
```
?

That would help it to specialize for the NamedTuple, which may help `merge` instability be a non-issue. Details [here in the Julia docs](https://docs.julialang.org/en/v1/manual/performance-tips/#Be-aware-of-when-Julia-avoids-specializing).

The tricky thing with all of this is that the compiler makes so few guarantees. Type stability is very context-dependent, even depending on what's been compiled so far. So code that seems fine when I write it can have subtle problems downstream.

Sorry this comment is so long. As might be obvious, I think there are lots of open questions around this kind of thing.